All tests pass with no code submitted, in x86-64 Assembly - Triangle

In the browser code editor, all tests pass if you do not submit any code. [This](https://exercism.org/tracks/x86-64-assembly/exercises/triangle/solutions/adep3zjh) user did just that and it has passed. 

I have not tested this with a CLI submission.

> I have not tested this with a CLI submission.

An empty solution does *not* pass the tests locally, at least not in the following environment:

    glibc version: 2.36
    gcc (Debian 12.2.0-14) 12.2.0
    GNU Make 4.3

    $ make
    triangle_test.c:24:test_equilateral_triangle_all_sides_are_equal:FAIL: Expected TRUE Was FALSE
    triangle_test.c:138:test_equilateral_triangle_any_side_is_unequal:PASS
    triangle_test.c:139:test_equilateral_triangle_no_sides_are_equal:PASS
    triangle_test.c:140:test_equilateral_triangle_all_zero_sides_is_not_a_triangle:PASS
    triangle_test.c:48:test_equilateral_triangle_sides_may_be_floats:FAIL: Expected TRUE Was FALSE
    triangle_test.c:54:test_isosceles_triangle_last_two_sides_are_equal:FAIL: Expected TRUE Was FALSE
    triangle_test.c:60:test_isosceles_triangle_first_two_sides_are_equal:FAIL: Expected TRUE Was FALSE
    triangle_test.c:66:test_isosceles_triangle_first_and_last_sides_are_equal:FAIL: Expected TRUE Was FALSE
    triangle_test.c:72:test_isosceles_triangle_equilateral_triangles_are_also_isosceles:FAIL: Expected TRUE Was FALSE
    triangle_test.c:146:test_isosceles_triangle_no_sides_are_equal:PASS
    triangle_test.c:147:test_isosceles_triangle_first_triangle_inequality_violation:PASS
    triangle_test.c:148:test_isosceles_triangle_second_triangle_inequality_violation:PASS
    triangle_test.c:149:test_isosceles_triangle_third_triangle_inequality_violation:PASS
    triangle_test.c:102:test_isosceles_triangle_sides_may_be_floats:FAIL: Expected TRUE Was FALSE
    triangle_test.c:108:test_scalene_triangle_no_sides_are_equal:FAIL: Expected TRUE Was FALSE
    triangle_test.c:152:test_scalene_triangle_all_sides_are_equal:PASS
    triangle_test.c:153:test_scalene_triangle_two_sides_are_equal:PASS
    triangle_test.c:154:test_scalene_triangle_may_not_violate_triangle_inequality:PASS
    triangle_test.c:155:test_scalene_triangle_sides_may_be_floats:PASS

    -----------------------
    19 Tests 8 Failures 0 Ignored
    FAIL
    make: *** [Makefile:29: all] Error 8


The real problem is the test runner ignores the exit code of the `make` task: https://github.com/exercism/x86-64-assembly-test-runner/blob/2360f54b486d33974faa1dbe96b6811f7ea270c9/run.sh#L13

The shell only gets the exit code returned by the `python3` command. As long as it runs the script successfully (and it always will), the exit code is 0 *regardless* of what "results.out" actually contains. It's an easy enough oversight to make; the [Scheme track] is currently suffering from a similar one.

A simple fix would be storing the exit code of `stdbuf` (which is non-zero when a test fails). Then, *after* running the Python script, exit with whatever `stdbuf` returned, e.g.

~~~sh
stdbuf -oL make > "$3"results.out 2>&1
TEST_RESULT=$(echo $?)
python3 "${cwd}"/process_results.py "$3"results.out
exit $TEST_RESULT
~~~

[Scheme track]: https://github.com/exercism/scheme/pull/285

There's more going on here, but at least it's certain that the test runner is at fault.

First, the musl libc that ships with Alpine really does appear to be initializing the argument registers with phantom data as @sudhackar suggested. Adding a few debug messages to the runner's `run.sh` script reveals as much:

<details>
<summary>Alpine test runner output:</summary>
<pre>
<code>
$ ./docker-run.sh triangle ../exercism-x86-64-assembly/exercises/practice/triangle
[+] Building 17.5s (11/11) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                             0.0s
 => => transferring dockerfile: 250B                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                0.0s
 => => transferring context: 113B                                                                                                                                0.0s
 => [internal] load metadata for docker.io/library/alpine:3.10                                                                                                   0.8s
 => [1/6] FROM docker.io/library/alpine:3.10@sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98                                             0.8s
 => => resolve docker.io/library/alpine:3.10@sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98                                             0.0s
 => => sha256:396c31837116ac290458afcb928f68b6cc1c7bdd6963fc72f52f365a2a89c1b5 2.80MB / 2.80MB                                                                   0.3s
 => => sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98 1.64kB / 1.64kB                                                                   0.0s
 => => sha256:e515aad2ed234a5072c4d2ef86a1cb77d5bfe4b11aa865d9214875734c4eeb3c 528B / 528B                                                                       0.0s
 => => sha256:e7b300aee9f9bf3433d32bc9305bfdd22183beb59d933b48d77ab56ba53a197a 1.47kB / 1.47kB                                                                   0.0s
 => => extracting sha256:396c31837116ac290458afcb928f68b6cc1c7bdd6963fc72f52f365a2a89c1b5                                                                        0.3s
 => [internal] load build context                                                                                                                                0.0s
 => => transferring context: 780B                                                                                                                                0.0s
 => [2/6] RUN apk add --no-cache coreutils gcc libc-dev make nasm python3                                                                                        7.9s
 => [3/6] WORKDIR /opt/test-runner                                                                                                                               0.1s
 => [4/6] COPY run.sh bin/                                                                                                                                       0.1s
 => [5/6] COPY process_results.py ./                                                                                                                             0.1s
 => [6/6] COPY debug.asm ./                                                                                                                                      0.1s
 => exporting to image                                                                                                                                           7.6s
 => => exporting layers                                                                                                                                          7.6s
 => => writing image sha256:c9efdc690ab9ca82ab7e12f2475f118b3b9bd9e67b34e4866f17d51dfa423837                                                                     0.0s
 => => naming to docker.io/library/test-runner                                                                                                                   0.0s
Solution code:                                                                                                                                                        
section .text

global is_equilateral
is_equilateral:
    ; Provide your implementation here
    ret

global is_isosceles
is_isosceles:
    ; Provide your implementation here
    ret

global is_scalene
is_scalene:
    ; Provide your implementation here
    ret

%ifidn __OUTPUT_FORMAT__,elf64
section .note.GNU-stack noalloc noexec nowrite progbits
%endif

triangle_test.c:137:test_equilateral_triangle_all_sides_are_equal:PASS
triangle_test.c:138:test_equilateral_triangle_any_side_is_unequal:PASS
triangle_test.c:139:test_equilateral_triangle_no_sides_are_equal:PASS
triangle_test.c:140:test_equilateral_triangle_all_zero_sides_is_not_a_triangle:PASS
triangle_test.c:141:test_equilateral_triangle_sides_may_be_floats:PASS
triangle_test.c:142:test_isosceles_triangle_last_two_sides_are_equal:PASS
triangle_test.c:143:test_isosceles_triangle_first_two_sides_are_equal:PASS
triangle_test.c:144:test_isosceles_triangle_first_and_last_sides_are_equal:PASS
triangle_test.c:145:test_isosceles_triangle_equilateral_triangles_are_also_isosceles:PASS
triangle_test.c:146:test_isosceles_triangle_no_sides_are_equal:PASS
triangle_test.c:147:test_isosceles_triangle_first_triangle_inequality_violation:PASS
triangle_test.c:148:test_isosceles_triangle_second_triangle_inequality_violation:PASS
triangle_test.c:149:test_isosceles_triangle_third_triangle_inequality_violation:PASS
triangle_test.c:150:test_isosceles_triangle_sides_may_be_floats:PASS
triangle_test.c:151:test_scalene_triangle_no_sides_are_equal:PASS
triangle_test.c:152:test_scalene_triangle_all_sides_are_equal:PASS
triangle_test.c:153:test_scalene_triangle_two_sides_are_equal:PASS
triangle_test.c:154:test_scalene_triangle_may_not_violate_triangle_inequality:PASS
triangle_test.c:155:test_scalene_triangle_sides_may_be_floats:PASS

 - - - - - - - - - - -  - - 
19 Tests 0 Failures 0 Ignored 
OK
stdbuf -oL make > "$3"results.out 2>&1 exited with code 0
Finishing with code 0
</code>
</details>

Now, replace Alpine with a stock Debian base image, and suddenly the `debug.asm` library does not link anymore:

    /usr/bin/ld: warning: debug.o: missing .note.GNU-stack section implies executable stack
    /usr/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
    collect2: error: ld returned 1 exit status
    make: *** [Makefile:32: tests] Error 1

Add the boilerplate assembler directive to `debug.asm` as needed, and now the tests clearly fail on an empty solution. The exit code is still 0, so the last step would be revising the runner script as I recommended above:

<details>
<summary>Debian test runner output:</summary>
<pre>
<code>
$ bash docker-run.sh triangle ../exercism-x86-64-assembly/exercises/practice/triangle
[+] Building 0.5s (11/11) FINISHED
 => [internal] load .dockerignore                                                                                                                                0.0s
 => => transferring context: 113B                                                                                                                                0.0s
 => [internal] load build definition from Dockerfile                                                                                                             0.0s
 => => transferring dockerfile: 451B                                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/debian:stable-slim                                                                                            0.3s
 => [1/6] FROM docker.io/library/debian:stable-slim@sha256:b09f68bffcf9c14f3105f262e92321d05abaf48460d1f43f884325bcd4395b95                                      0.0s
 => [internal] load build context                                                                                                                                0.0s
 => => transferring context: 4.29kB                                                                                                                              0.0s
 => CACHED [2/6] RUN apt update     && apt -qqy --no-install-recommends install         coreutils         gcc         libc-dev         make         nasm         0.0s
 => CACHED [3/6] WORKDIR /opt/test-runner                                                                                                                        0.0s
 => CACHED [4/6] COPY run.sh bin/                                                                                                                                0.0s
 => CACHED [5/6] COPY process_results.py ./                                                                                                                      0.0s
 => [6/6] COPY debug.asm ./                                                                                                                                      0.1s
 => exporting to image                                                                                                                                           0.1s
 => => exporting layers                                                                                                                                          0.0s
 => => writing image sha256:694e8c844dc66640846e344abeff35c54e64366a8a00cfe4a31222b7ff2c181e                                                                     0.0s
 => => naming to docker.io/library/test-runner                                                                                                                   0.0s
Solution code:
section .text

global is_equilateral
is_equilateral:
    ; Provide your implementation here
    ret

global is_isosceles
is_isosceles:
    ; Provide your implementation here
    ret

global is_scalene
is_scalene:
    ; Provide your implementation here
    ret

%ifidn __OUTPUT_FORMAT__,elf64
section .note.GNU-stack noalloc noexec nowrite progbits
%endif

triangle_test.c:24:test_equilateral_triangle_all_sides_are_equal:FAIL: Expected TRUE Was FALSE
triangle_test.c:138:test_equilateral_triangle_any_side_is_unequal:PASS
triangle_test.c:139:test_equilateral_triangle_no_sides_are_equal:PASS
triangle_test.c:140:test_equilateral_triangle_all_zero_sides_is_not_a_triangle:PASS
triangle_test.c:48:test_equilateral_triangle_sides_may_be_floats:FAIL: Expected TRUE Was FALSE
triangle_test.c:54:test_isosceles_triangle_last_two_sides_are_equal:FAIL: Expected TRUE Was FALSE
triangle_test.c:60:test_isosceles_triangle_first_two_sides_are_equal:FAIL: Expected TRUE Was FALSE
triangle_test.c:66:test_isosceles_triangle_first_and_last_sides_are_equal:FAIL: Expected TRUE Was FALSE
triangle_test.c:72:test_isosceles_triangle_equilateral_triangles_are_also_isosceles:FAIL: Expected TRUE Was FALSE
triangle_test.c:146:test_isosceles_triangle_no_sides_are_equal:PASS
triangle_test.c:147:test_isosceles_triangle_first_triangle_inequality_violation:PASS
triangle_test.c:148:test_isosceles_triangle_second_triangle_inequality_violation:PASS
triangle_test.c:149:test_isosceles_triangle_third_triangle_inequality_violation:PASS
triangle_test.c:102:test_isosceles_triangle_sides_may_be_floats:FAIL: Expected TRUE Was FALSE
triangle_test.c:108:test_scalene_triangle_no_sides_are_equal:FAIL: Expected TRUE Was FALSE
triangle_test.c:152:test_scalene_triangle_all_sides_are_equal:PASS
triangle_test.c:153:test_scalene_triangle_two_sides_are_equal:PASS
triangle_test.c:154:test_scalene_triangle_may_not_violate_triangle_inequality:PASS
triangle_test.c:155:test_scalene_triangle_sides_may_be_floats:PASS

-----------------------
19 Tests 8 Failures 0 Ignored
FAIL
make: *** [Makefile:29: all] Error 8
stdbuf -oL make > "$3"results.out 2>&1 exited with code 2
Finishing with code 0

</code>
</pre>
</details>

--- a/exercises/practice/triangle/.meta/example.asm
+++ b/exercises/practice/triangle/.meta/example.asm
@@ -49,6 +49,7 @@ global is_isosceles:
 is_isosceles:
         xor     eax, eax
         xor     ecx, ecx
+        xor     edx, edx
         movsd   xmm5, QWORD [rsp+8]
         movsd   xmm6, QWORD [rsp+16]
         movsd   xmm4, QWORD [rsp+24]

--- a/exercises/practice/triangle/.meta/example.asm
+++ b/exercises/practice/triangle/.meta/example.asm
@@ -21,6 +21,7 @@ triangle_equality:
 
 global is_equilateral:
 is_equilateral:
+        xor     eax, eax
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]
@@ -45,6 +46,8 @@ is_equilateral:
         ret
 global is_isosceles:
 is_isosceles:
+        xor     eax, eax
+        xor     ecx, ecx
         movsd   xmm5, QWORD [rsp+8]
         movsd   xmm6, QWORD [rsp+16]
         movsd   xmm4, QWORD [rsp+24]
@@ -72,6 +75,7 @@ is_isosceles:
 
 global is_scalene
 is_scalene:
+        xor     eax, eax
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]

--- a/exercises/practice/triangle/.meta/example.asm
+++ b/exercises/practice/triangle/.meta/example.asm
@@ -22,6 +22,7 @@ triangle_equality:
 global is_equilateral:
 is_equilateral:
         xor     eax, eax
+        xor     edx, edx
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]
         push    QWORD [rsp+24]

--- a/exercises/practice/triangle/triangle_test.c
+++ b/exercises/practice/triangle/triangle_test.c
@@ -1,6 +1,5 @@
 // Version: 0
 
-#include <stdbool.h>
 #include "vendor/unity.h"
 
 typedef struct {
@@ -9,9 +8,9 @@ typedef struct {
    double c;
 } triangle_t;
 
-extern bool is_equilateral(triangle_t sides);
-extern bool is_isosceles(triangle_t sides);
-extern bool is_scalene(triangle_t sides);
+extern int is_equilateral(triangle_t sides);
+extern int is_isosceles(triangle_t sides);
+extern int is_scalene(triangle_t sides);
 
 void setUp(void) {
 }
@@ -21,115 +20,115 @@ void tearDown(void) {
 
 void test_equilateral_triangle_all_sides_are_equal(void) {
     triangle_t sides = { 2, 2, 2 } ;
-    TEST_ASSERT_TRUE(is_equilateral(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_equilateral(sides));
 }
 
 void test_equilateral_triangle_any_side_is_unequal(void) {
     TEST_IGNORE();
     triangle_t sides = { 2, 3, 2 } ;
-    TEST_ASSERT_FALSE(is_equilateral(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_equilateral(sides));
 }
 
 void test_equilateral_triangle_no_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 5, 4, 6 } ;
-    TEST_ASSERT_FALSE(is_equilateral(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_equilateral(sides));
 }
 
 void test_equilateral_triangle_all_zero_sides_is_not_a_triangle(void) {
     TEST_IGNORE();
     triangle_t sides = { 0, 0, 0 } ;
-    TEST_ASSERT_FALSE(is_equilateral(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_equilateral(sides));
 }
 
 void test_equilateral_triangle_sides_may_be_floats(void) {
     TEST_IGNORE();
     triangle_t sides = { 0.5, 0.5, 0.5 } ;
-    TEST_ASSERT_TRUE(is_equilateral(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_equilateral(sides));
 }
 
 void test_isosceles_triangle_last_two_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 3, 4, 4 } ;
-    TEST_ASSERT_TRUE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_first_two_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 4, 4, 3 } ;
-    TEST_ASSERT_TRUE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_first_and_last_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 4, 3, 4 } ;
-    TEST_ASSERT_TRUE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_equilateral_triangles_are_also_isosceles(void) {
     TEST_IGNORE();
     triangle_t sides = { 4, 4, 4 } ;
-    TEST_ASSERT_TRUE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_no_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 2, 3, 4 } ;
-    TEST_ASSERT_FALSE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_first_triangle_inequality_violation(void) {
     TEST_IGNORE();
     triangle_t sides = { 1, 1, 3 } ;
-    TEST_ASSERT_FALSE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_second_triangle_inequality_violation(void) {
     TEST_IGNORE();
     triangle_t sides = { 1, 3, 1 } ;
-    TEST_ASSERT_FALSE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_third_triangle_inequality_violation(void) {
     TEST_IGNORE();
     triangle_t sides = { 3, 1, 1 } ;
-    TEST_ASSERT_FALSE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_isosceles(sides));
 }
 
 void test_isosceles_triangle_sides_may_be_floats(void) {
     TEST_IGNORE();
     triangle_t sides = { 0.5, 0.4, 0.5 } ;
-    TEST_ASSERT_TRUE(is_isosceles(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_isosceles(sides));
 }
 
 void test_scalene_triangle_no_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 5, 4, 6 } ;
-    TEST_ASSERT_TRUE(is_scalene(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_scalene(sides));
 }
 
 void test_scalene_triangle_all_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 4, 4, 4 } ;
-    TEST_ASSERT_FALSE(is_scalene(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_scalene(sides));
 }
 
 void test_scalene_triangle_two_sides_are_equal(void) {
     TEST_IGNORE();
     triangle_t sides = { 4, 4, 3 } ;
-    TEST_ASSERT_FALSE(is_scalene(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_scalene(sides));
 }
 
 void test_scalene_triangle_may_not_violate_triangle_inequality(void) {
     TEST_IGNORE();
     triangle_t sides = { 7, 3, 2 } ;
-    TEST_ASSERT_FALSE(is_scalene(sides));
+    TEST_ASSERT_EQUAL_INT(0, is_scalene(sides));
 }
 
 void test_scalene_triangle_sides_may_be_floats(void) {
     TEST_IGNORE();
     triangle_t sides = { 0.5, 0.4, 0.6 } ;
-    TEST_ASSERT_TRUE(is_scalene(sides));
+    TEST_ASSERT_EQUAL_INT(1, is_scalene(sides));
 }
 
 int main(void) {

